File: C++/count-good-integers-on-a-grid-path.cpp
// Time:  O(16 * 2 * 10 * 10)
// Space: O(16 + 2 * 10)

// dp
class Solution {
public:
    long long countGoodIntegersOnPath(long long l, long long r, string directions) {
        static const int L = 16;

        vector<bool> lookup(L);
        const auto& count = [&](int64_t n) {
            vector<int> digits(L);
            for (int i = L - 1; i >= 0; --i) {
                digits[i] = n % 10;
                n /= 10;
            }
            int64_t dp[2][10] = {};
            dp[1][0] = 1;
            for (int i = 0; i < L; ++i) {
                int64_t new_dp[2][10] = {};
                for (int t = 0; t < 2; ++t) {
                    const auto& bound = t ? digits[i] : 9;
                    for (int k = 0; k < 10; ++k) {
                        if (!dp[t][k]) {
                            continue;
                        }
                        for (int d = 0; d <= bound; ++d) {
                            int nk = k;
                            if (lookup[i]) {
                                if (d < k) {
                                    continue;
                                }
                                nk = d;
                            }
                            new_dp[t && d == digits[i]][nk] += dp[t][k];
                        }
                    }
                }
                for (int t = 0; t < 2; ++t) {
                    for (int k = 0; k < 10; ++k) {
                        dp[t][k] = new_dp[t][k];
                    }
                }
            }
            int64_t result = 0;
            for (int t = 0; t < 2; ++t) {
                for (int k = 0; k < 10; ++k) {
                    result += dp[t][k];
                }
            }
            return result;
        };

        int i = 0, j = 0;
        lookup[i * 4 + j] = true;
        for (const auto& x : directions) {
            if (x == 'D') {
                ++i;
            } else {
                ++j;
            }
            lookup[i * 4 + j] = true;
        }
        return count(r) - count(l - 1);
    }
};

// Time:  O(16 * 2 * 10 * 10)
// Space: O(16 + 2 * 10)
// dp
class Solution2 {
public:
    long long countGoodIntegersOnPath(long long l, long long r, string directions) {
        static const int L = 16;

        vector<bool> lookup(L);
        const auto& count = [&](int64_t n) {
            vector<int> digits(L);
            for (int i = L - 1; i >= 0; --i) {
                digits[i] = n % 10;
                n /= 10;
            }
            vector<vector<int64_t>> dp(2, vector<int64_t>(10));
            dp[1][0] = 1;
            for (int i = 0; i < L; ++i) {
                vector<vector<int64_t>> new_dp(2, vector<int64_t>(10));
                for (int t = 0; t < 2; ++t) {
                    const auto& bound = t ? digits[i] : 9;
                    for (int k = 0; k < 10; ++k) {
                        if (!dp[t][k]) {
                            continue;
                        }
                        for (int d = 0; d <= bound; ++d) {
                            int nk = k;
                            if (lookup[i]) {
                                if (d < k) {
                                    continue;
                                }
                                nk = d;
                            }
                            new_dp[t && d == digits[i]][nk] += dp[t][k];
                        }
                    }
                }
                dp = move(new_dp);
            }
            int64_t result = 0;
            for (const auto& row : dp) {
                result += accumulate(cbegin(row), cend(row), 0LL);
            }
            return result;
        };

        int i = 0, j = 0;
        lookup[i * 4 + j] = true;
        for (const auto& x : directions) {
            if (x == 'D') {
                ++i;
            } else {
                ++j;
            }
            lookup[i * 4 + j] = true;
        }
        return count(r) - count(l - 1);
    }
};

File: Python/count-good-integers-on-a-grid-path.py
# Time:  O(16 * 2 * 10 * 10)
# Space: O(16 + 2 * 10)

# dp
class Solution(object):
    def countGoodIntegersOnPath(self, l, r, directions):
        """
        :type l: int
        :type r: int
        :type directions: str
        :rtype: int
        """
        L = 16
        def count(n):
            digits = [0]*L
            for i in reversed(xrange(len(digits))):
                digits[i] = n%10
                n //= 10
            dp = [[0]*10 for _ in xrange(2)]
            dp[1][0] = 1
            for i in xrange(L):
                new_dp = [[0]*10 for _ in xrange(2)]
                for t in xrange(2):
                    bound = digits[i] if t else 9
                    for k in xrange(10):
                        if not dp[t][k]:
                            continue
                        for d in xrange(bound+1):
                            nk = k
                            if lookup[i]:
                                if d < k:
                                    continue
                                nk = d                                
                            new_dp[t and d == bound][nk] += dp[t][k]
                dp = new_dp
            return sum(sum(row) for row in dp)

        i = j = 0
        lookup = [False]*L
        lookup[i*4+j] = True
        for x in directions:
            if x == 'D':
                i += 1
            else:
                j += 1
            lookup[i*4+j] = True
        return count(r)-count(l-1)


# Time:  O(16 * 2 * 10 * 10)
# Space: O(16 * 2 * 10)
# memoization
class Solution2(object):
    def countGoodIntegersOnPath(self, l, r, directions):
        """
        :type l: int
        :type r: int
        :type directions: str
        :rtype: int
        """
        L = 16
        def count(n):
            def memoization(i, t, k):
                if i == L:
                    return 1
                if memo[i][t][k] == -1:
                    memo[i][t][k] = 0
                    bound = digits[i] if t else 9
                    for d in xrange(bound+1):
                        nk = k
                        if lookup[i]:
                            if d < k:
                                continue
                            nk = d   
                        memo[i][t][k] += memoization(i+1, t and d == bound, nk)
                return memo[i][t][k]

            digits = [0]*L
            for i in reversed(xrange(len(digits))):
                digits[i] = n%10
                n //= 10
            memo = [[[-1]*10 for _ in xrange(2)] for _ in xrange(L)]
            return memoization(0, True, 0)

        i = j = 0
        lookup = [False]*L
        lookup[i*4+j] = True
        for x in directions:
            if x == 'D':
                i += 1
            else:
                j += 1
            lookup[i*4+j] = True
        return count(r)-count(l-1)
